oqpy/subroutines.py

```
@@ -66,97 +83,86 @@ def increment_variable(int[32] i) {
+    arguments = []  # used in the ast definition of the subroutine
+    for argname in argnames[1:]:  # arg 0 should be program
+        if argname not in type_hints:
+            raise ValueError(f"No type hint provided for {argname} on subroutine {name}.")
+        input_ = inputs[argname] = type_hints[argname](name=argname)
+
+        if isinstance(input_, _ClassicalVar):
+            arguments.append(ast.ClassicalArgument(input_.type, ast.Identifier(argname)))
+        elif isinstance(input_, Qubit):
+            arguments.append(ast.QuantumArgument(ast.Identifier(input_.name), None))
+        else:
+            raise ValueError(
+                f"Type hint for {argname} on subroutine {name} is not an oqpy variable type."
+            )
+
+    inner_prog = oqpy.Program()
+    for input_val in inputs.values():
+        inner_prog._mark_var_declared(input_val)
+    output = func(inner_prog, **inputs)
+    inner_prog.autodeclare()
+    inner_prog._state.finalize_if_clause()
+    body = inner_prog._state.body
+    if isinstance(output, OQPyExpression):
+        return_type = output.type
+        body.append(ast.ReturnStatement(to_ast(inner_prog, output)))
+    elif output is None:
+        return_type = None
+        if type_hints.get("return", False):
+            return_hint = type_hints["return"]()
+            if isinstance(return_hint, _ClassicalVar):
+                return_type = return_hint.type
+            elif return_hint is not None:
+                raise ValueError(
+                    f"Type hint for return variable on subroutine {name} is not an oqpy classical type."
+                )
+    else:
+        raise ValueError("Output type of subroutine {name} was neither oqpy expression nor None.")
+    stmt = ast.SubroutineDefinition(
+        identifier,
+        arguments=arguments,
+        return_type=return_type,
+        body=body,
+    )
+    stmt.annotations = make_annotations(annotations)
+
+    if prog is not None:
```

Maybe raise an error if prog is None and in_place_declaration is true?